### PR TITLE
Extract Bootstrap into its own section

### PR DESCRIPTION
**Pull request by jieyouxu**
Other changes:

- Add brief explanation for `Step` and `Builder::ensure` as core Bootstrap internal concepts.

Feedback, corrections, rewords and other contributions all very welcomed.

### CODE REVIEW

File: src/building/bootstrapping/common-commands.md
Comment (onur-ozkan) [inline src/building/bootstrapping/common-commands.md:1]: This page will require synchronization PRs quite frequently and people can easily see command references with `x --help`.

File: src/building/bootstrapping/how-bootstrap-does-it.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  for steps which are dependencies of other steps.
+
+## The entry points
+
+There's a couple of preliminary steps before core Bootstrap code is reached:
+
+1. Shell script: [`./x`](https://github.com/rust-lang/rust/blob/master/x) or [`./x.ps1`](https://github.com/rust-lang/rust/blob/master/x.ps1)
+2. Convenience wrapper script: [`x.py`](https://github.com/rust-lang/rust/blob/master/x.py)
+3. [`src/bootstrap/bootstrap.py`](https://github.com/rust-lang/rust/blob/master/src/bootstrap/bootstrap.py)
+4. [`src/bootstrap/src/bin/main.rs`](https://github.com/rust-lang/rust/blob/master/src/bootstrap/src/bin/main.rs)
```
Comment (onur-ozkan) [inline src/building/bootstrapping/how-bootstrap-does-it.md:45]: We also support `make` as an entrypoint https://github.com/rust-lang/rust/blob/master/INSTALL.md#configure-and-make

**Review by onur-ozkan — COMMENTED** [review]
thank you for doing this!

File: src/building/bootstrapping/common-commands.md
Comment (jieyouxu) [reply src/building/bootstrapping/common-commands.md:1]: I'll drop this page.

File: src/building/bootstrapping/how-bootstrap-does-it.md (hunk 1 above)
Comment (jieyouxu) [reply src/building/bootstrapping/how-bootstrap-does-it.md]: Thanks for the catch, added

**Review by onur-ozkan — APPROVED** [review]

**Review by spastorino — APPROVED** [review]

**Pull request merged**